python/dolfinx/fem/petsc.py

```diff
@@ -182,30 +241,96 @@ def create_matrix(
 
 
 @functools.singledispatch
-def assemble_vector(L: typing.Any, constants=None, coeffs=None) -> PETSc.Vec:
-    """Assemble linear form into a new PETSc vector.
+def assemble_vector(
+    L: typing.Union[Form, Iterable[Form]],
+    constants: typing.Optional[npt.NDArray, Iterable[npt.NDArray]] = None,
+    coeffs: typing.Optional[
+        typing.Union[
+            dict[tuple[IntegralType, int], npt.NDArray],
+            Iterable[dict[tuple[IntegralType, int], npt.NDArray]],
+        ]
+    ] = None,
+    kind: typing.Optional[str] = None,
+) -> PETSc.Vec:
+    """Assemble linear form(s) into a new PETSc vector.
+
+    Three cases are supported:
+
+    1. If ``L`` is a single linear form, the form is assembled into a
+       ghosted PETSc vector.
+
+    2. If ``L`` is a sequence of linear forms and ``kind`` is ``None``
+       or is ``PETSc.Vec.Type.MPI``, the forms are assembled into a
+       vector ``b`` such that ``b = [b_0, b_1, ..., b_n, b_0g, b_1g,
+       ..., b_ng]`` where ``b_i`` are the entries associated with the
+       'owned' degrees-of-freedom for ``L[i]`` and ``b_ig`` are the
+       'unowned' (ghost) entries for ``L[i]``.
+
+       For this case, the returned vector has an attribute ``_blocks``
+       that holds the local offsets into ``b`` for the (i) owned and
+       (ii) ghost entries for each ``L[i]``. See :func:`create_vector`
+       for a description of the offset blocks.
+
+    3. If ``L`` is a sequence of linear forms and ``kind`` is
+       ``PETSc.Vec.Type.NEST``, the forms are assembled into a PETSc
+       nested vector ``b`` (a nest of ghosted PETSc vectors) such that
+       ``L[i]`` is assembled into into the ith nested matrix in ``b``.
+
+    Constants and coefficients that appear in the forms(s) can be passed
+    to avoid re-computation of constants and coefficients. The functions
```

Comment:
re-computation -> re-packing (consistent with terminology of function names, makes link with functions mentioned below clearer)